Fixes and simplifications for `.ramified_primes()`, `.discriminant()` and `.is_isomorphic` of quaternion algebras

S17A05 opened the pull request:
1. Removed unnecessary product and factorization for `.ramified_primes()`
2. Adapted `is_isomorphic` to reduce unnecessary calculations
3. Fixed a bug in `.discriminant()` and `.ramified_primes()` where rational invariants caused errors
4. Removed `.hilbert_conductor` from `sage.arith.misc` import list and added `.hilbert_symbol`
5. Reduced restriction of `.ramified_primes()` to rational quaternion algebras in docstring in preparation for planned `.ramified_places()` function over number fields (currently being worked on)

In more detail:
1. The original workflow for `.ramified_primes()` went as follows:
    - Call `.discriminant()`, which calls `.hilbert_conductor`
    - Inside `.hilbert_conductor`, the ramified primes are computed and their product (the discriminant of the quaternion algebra) is returned
    - Finally, factor the discriminant back into its prime factors
    
    Hence we have a redundant product and, more crucially, a redundant prime factorization. This fix modifies `.ramified_primes()` to instead directly build the list computed in `.hilbert_conductor()` (up to a bug fix described in 3.) and return it; the list might not always be sorted by magnitude of primes, so an optional argument `sorted` (set to `False` by default) allows to enforce this (small to large) sorting. Furthermore, `.discriminant()` has been adapted to directly take the product of the list returned by `.ramified_primes()` (only in the rational case, for now - see 5.)

2. Since the `.discriminant()`-function needs to compute all (finite) ramified primes (this was also true before this PR, it was just hidden inside `.hilbert_conductor()` instead), the function `.is_isomorphic()` now compares the unsorted lists of finite ramified primes to decide whether two rational quaternion algebras are isomorphic.

3. The function `sage.arith.misc.hilbert_conductor` requires its arguments to be integers (to create certain lists of prime divisors); since it was originally used to determine the discriminant (and, as explained in 1., the ramified primes), it raises an error when the invariants are proper rational numbers. To get around the analogous error for the method `.hilbert_symbol`, we instead look at the numerators and denominators of both invariants separately, using the fact that we can (purely on a mathematical level) rescale both invariants by the squares of their respective denominators without leaving the isomorphism class of the algebra.

4. The only call to `sage.arith.misc.hilbert_conductor` in quaternion_algebra.py was given in the old computation of the discriminant (the other `.hilbert_conductor` in the code, also in `.discriminant()`, refers to the one in `sage.rings.number_field`), so it was removed from the import list. The new approach to `.ramified_primes()` requires `sage.arith.misc.hilbert_symbol`, which was added to the import list.

5. As of now the `.ramified_primes()`-method is only supported for rational quaternion algebras. I'm currently working on a version over number fields, but once it works correctly this will be implemented as a new function `.ramified_places` (Update: see #37173) ~~to distinguish between different formats (prime numbers vs ideals) over $\mathbb{Q}$~~ (Update: this wasn't really feasible, see the issues discussed in #7596; thanks to @yyyyx4 for pointing me towards this discussion) ~~and, furthermore,~~ to not cause confusion using the term "primes" for the Archimedean real places where a quaternion algebra might ramify. Hence the implementation restriction in the docstring of `.ramified_primes()` was removed, but the method still throws a ValueError if not called with a quaternion algebra defined over the rational numbers.

#sd123

S17A05 committed:
Various fixes to ramified_primes and discriminant

- Fixed a bug in .discriminant() and .ramified_primes() for rational invariants
- Removed unnecessary product and factoring for .ramified_primes()
- Reduced description of .ramified_primes() to rational quaternion algebras in preparation of planned .ramified_places() methods over number fields
(Amended: Fixed trailing whitespaces)

grhkm21 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -1058,24 +1056,47 @@ def discriminant(self):
+        else:
+            return sorted([p for p in set([2]).union(prime_divisors(self._a.numerator()),
+                prime_divisors(self._a.denominator()), prime_divisors(self._b.numerator()),
+                prime_divisors(self._b.denominator())) if hilbert_symbol(self._a, self._b, p) == -1])
```
The code will be nicer if you first compute the list, then sorted it if `sorted` is set. So
```python
result = [...]
if sorted: result.sort()
return result
```

S17A05 replied:
Thanks for the suggestion, I'll simplify it that way

grhkm21 commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -1058,24 +1056,47 @@ def discriminant(self):
+            return prod(self.ramified_primes())
 
-    def ramified_primes(self):
+    def ramified_primes(self, sorted=False):
```
Also, I suppose this should be cached; simply add `@cached_method` above.

grhkm21 replied:
One more comment: is there any use for `sorted=False` at all, apart from perhaps a tiny speedup?

S17A05 replied:
Good point! Regarding sorted: It's really only there for speeding up a bit, since e.g. the computation of the discriminant does not need the ordering and I believe that the comparison in `.is_isomorphic` should also be faster by comparing the lists as sets than by sorting them first.

S17A05 replied:
> Good point! Regarding sorted: It's really only there for speeding up a bit, since e.g. the computation of the discriminant does not need the ordering and I believe that the comparison in `.is_isomorphic` should also be faster by comparing the lists as sets than by sorting them first.

Correction to the above, I remembered that I also wanted to put it in there to easily mimic the behavior of the old variant, which correctly sorted the primes during factorization. The reason I still decided to default to `False` was for speed, though.

grhkm21 replied:
If so, I recommend removing the option. It makes the output more clear to the user, and the time it takes to sort (I assume) < 100 elements is negligible anyways, compared to say the hilbert symbols.

S17A05 replied:
That's a fair point, I'll remove it (and adapt the isomorphism check accordingly)

S17A05 committed:
Implemented reviewer recommendations

- Slightly modified description of `.discriminant()` to be more accurate
- Cached the method `.ramified_primes()`
- Removed optional parameter `sorted` in `.ramified_primes()`, sorting the primes by default instead.
- Removed `set()` functions in `.is_isomorphic` since the lists can be compared directly, as they are always sorted according to the above change.
- Made description of `.is_isomorphic` slightly more explicit.

S17A05 committed:
Build fixes

- Added `.prime_divisors` to import list
- Modified product call in `.discriminant()`
